src/compiler/checker.ts

```diff
@@ -573,7 +573,16 @@ namespace ts {
                         declarationNameToString(propertyName), typeof nameArg === "string" ? nameArg : declarationNameToString(nameArg));
                     return undefined;
                 }
-                if (result.flags & SymbolFlags.BlockScopedVariable) {
+
+                // Only check for block-scoped variable if we are looking for the name with variable meaning
+                // For example,
+                //      declare module foo {
+                //          interface bar {}
+                //      }
+                // let foo/*1*/: foo/*2*/.bar;
+                // The foo at /*1*/ and /*2*/ will share same symbol with two meaning block-scope variable and namespace module.
+                // However, only when we try to resolve name in /*1*/ which is used in variable position, we want to check for block- scoped

```

Comment:
I don't think we need this comment; we don't have comments anywhere else where we check the meaning.

Instead, I think a better change would be to assert that your result has the appropriate symbol before returning it below.

```TypeScript
if (symbol && !(symbol.flags & meaning)) {
    Debug.fail(`Symbol '${symbol.name}' with flags '${symbol.flags}' did not have meaning '${meaning}'.`);
}

return symbol;